### ISSUE TITLE
Document pre-requisites for running Elastic Agent in unprivileged mode

### ISSUE THREAD

**Issue by ycombinator** [elastic/elastic-agent#4705]
## Background
Traditionally, privileged users (e.g. `root` on Linux) run Elastic Agent on a host.  However, with https://github.com/elastic/elastic-agent/pull/3598, https://github.com/elastic/elastic-agent/pull/4362, https://github.com/elastic/elastic-agent/pull/4264, and other follow-up PRs, it is now possible to run Elastic Agent with an unprivileged user.

## Problem statement

Running Agent as an unprivileged user has consequences. Not only does the Agent itself run as an unprivileged user, but so do the process components it orchestrates, e.g. the various Beats. Consequently, any integrations being handled by such components, e.g. `system`, might not have the necessary access on the host to collect all the data they can when running as a privileged user. The result is that users do not see data they might be expecting in these integrations' dashboards. Some examples of this situation are:
* https://github.com/elastic/elastic-agent/issues/4647
* https://github.com/elastic/elastic-agent/issues/4653
* https://github.com/elastic/elastic-agent/issues/4675

Similarly, users might encounter other issues related to the installing or running of Elastic Agent in privileged mode.  Some examples of this situation are:
* https://github.com/elastic/elastic-agent/issues/4690
* https://github.com/elastic/elastic-agent/issues/4703

## Definition of done
Let's use this issue to collect any pre-requisites a user must perform to install and run Elastic Agent in unprivileged mode, as well as any other gotchas they might run into when using the `system` integration with an Elastic Agent running in unprivileged mode.

For each pre-requisite let's capture the following information:
1. What steps does the user need to take as a prerequisite to running Elastic Agent in unprivileged mode?
2. What would the impact be if these prerequisite steps were _not_ taken? Or, put differently, what functionality is enabled as a result of taking these prerequisite steps?
3. What symptoms (e.g. errors) will the user observe and where if these prerequisite steps were _not_ taken?

```[tasklist]
### MacOS
- [x] Try installing Elastic Agent in unprivileged mode.
- [x] Try installing Elastic Agent in unprivileged mode in a custom location using the `--base-path` option
- [x] Try running various Elastic Agent subcommands (e.g. `inspect`, `status`, etc.).
- [x] Try installing the `system` integration on an Elastic Agent in unprivileged mode and viewing its dashboards. Compare the results with installing the `system` integration on Elastic Agent in privileged mode.
```

```[tasklist]
### Linux
- [x] Try installing Elastic Agent in unprivileged mode.
- [x] Try installing Elastic Agent in unprivileged mode in a custom location using the `--base-path` option
- [x] Try running various Elastic Agent subcommands (e.g. `inspect`, `status`, etc.).
- [x] Try installing the `system` integration on an Elastic Agent in unprivileged mode and viewing its dashboards. Compare the results with installing the `system` integration on Elastic Agent in privileged mode.
```

```[tasklist]
### Windows
- [x] Try installing Elastic Agent in unprivileged mode.
- [x] Try installing Elastic Agent in unprivileged mode in a custom location using the `--base-path` option
- [x] Try running various Elastic Agent subcommands (e.g. `inspect`, `status`, etc.).
- [x] Try installing the `system` integration on an Elastic Agent in unprivileged mode and viewing its dashboards. Compare the results with installing the `system` integration on Elastic Agent in privileged mode.
```

**Comment by ycombinator** [comment]
cc: @kilfoyle

**Comment by ycombinator** [comment]
In https://github.com/elastic/elastic-agent/issues/4125#issuecomment-2134015269, @kilfoyle said:

> @kaanyalti I think the "pre-requisites and gotchas" could go in [tables like these](https://ingest-docs_bk_1087.docs-preview.app.elstc.co/guide/en/fleet/master/elastic-agent-unprivileged.html#unprivileged-command-behaviors), but we can update the format once the list becomes more clear.

**Comment by kaanyalti** [comment]
# Mac tests
 - agent version 8.14.0
- Production ESS 8.14
- Create agent policy without system integration
- Enroll in fleet in unprivileged mode
- logs
![Image](https://github.com/elastic/elastic-agent/assets/23665523/1a5c8ccd-b922-4689-bb8b-5d7fcf0e26cd)
```
[elastic_agent.filebeat][error] Harvester could not be started on new file: /var/log/system.log, Err: error setting up harvester: Harvester setup failed. Unexpected file opening error: Failed opening /var/log/system.log: open /var/log/system.log: permission denied
```
- metrics
![Image](https://github.com/elastic/elastic-agent/assets/23665523/a1d03172-b95b-4b2d-961e-15b9d8e8c35a)

```  
[elastic_agent.metricbeat][error] add_cloud_metadata: received error failed requesting openstack metadata: Get "https://169.254.169.254/2009-04-04/meta-data/instance-id": dial tcp 169.254.169.254:443: i/o timeout
```
found this open issue related to these logs and metrics https://github.com/elastic/beats/issues/39733
## Testing commands:
- Failed `enroll`
	- Installed fleet managed agent
	- Unenrolled the agent
	- Tried to enroll the agent
	- Got the following error right away
```
[Error: fail to enroll: acquiring metadata failed: failed to create new agent info: could not get agent info from store: fail to read configuration /Library/Elastic/Agent/fleet.enc for the agent: fail to decode bytes: cipher: message authentication failed
For help, please see our troubleshooting guide at https://www.elastic.co/guide/en/fleet/8.14/fleet-troubleshooting.html](<{"log.level":"info","@timestamp":"2024-06-06T22:05:59.706-0700","log.origin":{"file.name":"cmd/enroll_cmd.go","file.line":506},"message":"Starting enrollment to URL: https://a8ef7f060207437a961f91ef9c7b3f42.fleet.us-central1.gcp.cloud.es.io:443/","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2024-06-06T22:05:59.933-0700","log.origin":{"file.name":"cmd/enroll_cmd.go","file.line":512},"message":"1st enrollment attempt failed, retrying enrolling to URL: https://a8ef7f060207437a961f91ef9c7b3f42.fleet.us-central1.gcp.cloud.es.io:443/ with exponential backoff (init 1s, max 10s)","ecs.version":"1.6.0"}
Error: fail to enroll: acquiring metadata failed: failed to create new agent info: could not get agent info from store: fail to read configuration /Library/Elastic/Agent/fleet.enc for the agent: fail to decode bytes: cipher: message authentication failed
For help, please see our troubleshooting guide at https://www.elastic.co/guide/en/fleet/8.14/fleet-troubleshooting.html>)
```
- Failed `upgrade`
	- install 8.13.4 in unprivileged mode and then try to upgrade
	- couldn't install 8.13.4 in unprivileged mode, only supported in linux
	```
	Error: unable to perform install command, unprivileged is currently only supported on Linux
	```
	- So I re-installed 8.14.0 and called upgrade with version equal to 8.14.0 as well just to see what happens. 
		- I got the following output
		```	
		Error: Failed trigger upgrade of daemon: agent version is already 8.14.0 (hash: d0bce03b68012224ac1bab4ca20ba36060d135a3)
		```
		which probably means there is no issue with upgardes, but I cannot know for sure, going to test this for linux.
- Failed `inspect`
```
Error: error loading agent config: error loading raw config: fail to read configuration /Library/Elastic/Agent/fleet.enc for the elastic-agent: fail to decode bytes: cipher: message authentication failed
For help, please see our troubleshooting guide at https://www.elastic.co/guide/end/fleet/8.14/fleet-troubleshooting.html
```
- Failed `otel`
```
fileconsumer/file.go:260 Started watching file {"kind": "receiver", "name": "filelog", "data_type": "logs", "component": "fileconsumer", "path": "/var/log/system.log"ssh}  
1 error occurred:  

- could not load agent info: could not get agent info from store: fail to read configuration /Library/Elastic/Agent/fleet.enc for the agent: fail to decode bytes: cipher: message authentication failed
```

## Dashboards:
- [Logs System] Syslog dashboard
	- Privileged
![Image](https://github.com/elastic/elastic-agent/assets/23665523/251af725-cca2-4813-bd32-4479fed2a88c)
	- Unprivileged
![Image](https://github.com/elastic/elastic-agent/assets/23665523/64c41396-6e38-44b1-8850-73bfcddc197b)

- [Metrics System] Host overview
	- Priveleged
![Image](https://github.com/elastic/elastic-agent/assets/23665523/6c2f6556-06d6-4d57-ac3c-b1a0e65dbe3c)
	- Unprivileged
![Image](https://github.com/elastic/elastic-agent/assets/23665523/5d0f755f-7e76-47b5-806b-362a637ef8c6)
	Agent doesn't seem to have all the processes listed in the cpu and memory usage lists. I ran a cpu stress test, saw the cpu usage go up, but the list did not get updated
	
- [Elastic Agent] Agent Info
	- Privileged
![Image](https://github.com/elastic/elastic-agent/assets/23665523/8d885673-ef09-43c2-b0d0-2cd8160ca986)
	- Unprivileged
![Image](https://github.com/elastic/elastic-agent/assets/23665523/7530729f-6b56-4e9c-8cc7-d5fbff71af9f)
	
- [Elastic Agent] Integrations
	- Privileged
![Image](https://github.com/elastic/elastic-agent/assets/23665523/93aee8c9-2160-4531-9090-02ea185d523c)
	- Unprivileged
![Image](https://github.com/elastic/elastic-agent/assets/23665523/e4b1ac71-5ce2-4d56-8e15-a7936bd74740)




# Linux tests
- agent version 8.14.0
- Production ESS 8.14
- Use the same policy used for mac testing
- Install agent in fleet mode, unprivileged
- check the integration
- log
	- Again the same harvester error
```
[elastic_agent.filebeat][error] Harvester could not be started on new file: /var/log/auth.log.1, Err: error setting up harvester: Harvester setup failed. Unexpected file opening error: Failed opening /var/log/auth.log.1: open /var/log/auth.log.1: permission denied
```
- metrics
	- different error message failure
	```
	[elastic_agent.metricbeat][error] error getting filesystem usage for /run/user/1000/gvfs: error in Statfs syscall: permission denied
	```
## Testing commands:
- `component`
- Failed `enroll`
	- Installed fleet managed agent
	- Unenrolled the agent
	- Tried to enroll the agent
	- Agent enrollment got confirmed, but incoming data did not get confirmed
	- Agent status got stuck in `updating`
	- running `sudo elastic-agent status` gave the following error message
	```
	Error: failed to communicate with Elastic Agent daemon: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing: dial unix /opt/Elastic/Agent/elastic-agent.sock: connect: connection refused"
	For help, please see our troubleshooting guide at https://www.elastic.co/guide/en/fleet/8.14/fleet-troubleshooting.html
	```
- Note `upgrade`
	- Installed 8.13.4
	- ran `sudo elastic-agent upgrade 8.14.0`
	- Upgrade seemed to have worked, but got the following eror when checking the status 
	```
	┌─ fleet
	│  └─ status: (FAILED) status code: 500, fleet-server returned an error: BadRequest, message: failed to update upgrade_details: upgrade_details no action for id "" found
	├─ elastic-agent
	│  └─ status: (HEALTHY) Running
	└─ upgrade_details
	   ├─ target_version: 8.14.0
	   ├─ state: UPG_WATCHING
	   └─ metadata
	```
	- This is the same for privileged as well, so I may be doing something wrong. Going to try to upgrade through fleet UI
	- Root elastic agent upgrade through fleet ui worked
	- unprivileged agent upgrade through the fleet ui worked
## Dashboards:
- [Logs System] Syslog dashboard
	- privileged
![Image](https://github.com/elastic/elastic-agent/assets/23665523/63b4e2fe-0a56-45e2-95bf-c7957feef063)
	- unprivileged
![Image](https://github.com/elastic/elastic-agent/assets/23665523/75c2fb05-35ce-4c11-b49d-e6df4c3aa307)
	![[Pasted image 20240607113431.png]]
- [Metrics System] Host overview
	- privileged
![Image](https://github.com/elastic/elastic-agent/assets/23665523/dbde24d2-bd99-43e4-8054-e60c18dd710d)
	- unprivileged
![Image](https://github.com/elastic/elastic-agent/assets/23665523/3c6c1553-d982-43cd-abe5-cf0fa1bff254)
	- Here again the list of processes are different between the privileged and unprivileged agents
	
- [Elastic Agent] Agent Info
	- privileged
![Image](https://github.com/elastic/elastic-agent/assets/23665523/e2ee9f65-ef03-40a9-9c43-c2e6759637fa)
	- unprivileged
![Image](https://github.com/elastic/elastic-agent/assets/23665523/00790964-94fd-4155-bc87-af1d01808198)
- [Elastic Agent] Integrations 
	- privileged
![Image](https://github.com/elastic/elastic-agent/assets/23665523/4bf67197-692b-4517-a220-706748418f27)
	- unprivileged
![Image](https://github.com/elastic/elastic-agent/assets/23665523/91184fcf-9b6f-40a0-85e5-9fe53c96bed6)

**Comment by kaanyalti** [comment]
https://github.com/elastic/elastic-agent/issues/4887

**Comment by jlind23** [comment]
@ycombinator @pierrehilbert now that @VihasMakwana and @kaanyalti tested all the different combinations (unless I am missing something) what would be the next step here? Should @VihasMakwana start writing down all the settings that need to be changed when running unprivileged?

**Comment by VihasMakwana** [comment]
# Windows

## Dashboards after giving sufficient permissions as per [this comment](https://github.com/elastic/elastic-agent/issues/4705#issuecomment-2180934241)
- ### [Elastic Agent] Overview
<pre> Unprivileged 
<img width="1247" alt="Screenshot 2024-06-20 at 7 33 51 PM" src="https://github.com/elastic/elastic-agent/assets/121151420/1490bee7-7367-47c0-b799-1b768d16c9fb">
 </pre>
<pre> Privileged
 <img width="1251" alt="Screenshot 2024-06-20 at 5 29 25 PM" src="https://github.com/elastic/elastic-agent/assets/121151420/612fb16a-a7bc-404a-9be9-20e1de55d0c7"> 
  </pre>

- ### [Elastic Agent] Agent metrics
<pre> Unprivileged
<img width="1249" alt="Screenshot 2024-06-20 at 8 06 13 PM" src="https://github.com/elastic/elastic-agent/assets/121151420/6b031e01-d636-47d9-b1af-fc0492d85eba">
 </pre>
<pre> Privileged 
<img width="1246" alt="Screenshot 2024-06-20 at 5 30 40 PM" src="https://github.com/elastic/elastic-agent/assets/121151420/3c7b36ce-b9fb-483a-a6f3-c7c860403d1b"> 
<img width="1501" alt="Screenshot 2024-06-20 at 5 30 50 PM" src="https://github.com/elastic/elastic-agent/assets/121151420/b35dac2d-e8d6-43ae-aee0-7ffd25942a80">
</pre>

- ### [Elastic Agent] Input Metrics
<pre> Unprivileged
<img width="841" alt="Screenshot 2024-06-20 at 8 07 24 PM" src="https://github.com/elastic/elastic-agent/assets/121151420/4f5c0bb3-88a0-4c3a-bee7-9847d7371414">
</pre>
<pre> Privileged 
<img width="849" alt="Screenshot 2024-06-20 at 5 31 46 PM" src="https://github.com/elastic/elastic-agent/assets/121151420/8d9890f7-3fa3-4f3b-91b5-e8760cfa9664">
 </pre>
 
- ### [Elastic Agent] Filestream Input Metrics
<pre> Unprivileged
<img width="1287" alt="Screenshot 2024-06-20 at 8 10 43 PM" src="https://github.com/elastic/elastic-agent/assets/121151420/763fa488-63e4-400d-ad10-b14ef2f8b968">
 </pre>
<pre> Privileged 
<img width="1281" alt="Screenshot 2024-06-20 at 5 32 34 PM" src="https://github.com/elastic/elastic-agent/assets/121151420/30d09482-d615-43a2-8058-f76fee1d08ec"> </pre>
 
- ### [Elastic Agent] Info
 <pre> Unprivileged
<img width="1250" alt="Screenshot 2024-06-20 at 8 13 27 PM" src="https://github.com/elastic/elastic-agent/assets/121151420/456febb5-0628-431b-a881-8c6eb9e8bd60">
 </pre>
 <pre> Privileged
<img width="1244" alt="Screenshot 2024-06-20 at 5 35 00 PM" src="https://github.com/elastic/elastic-agent/assets/121151420/052aeba0-2e38-4524-ac0a-fb6021c75c68">
 </pre>
 
- ### System Overview
 <pre> Unprivileged
<img width="1500" alt="Screenshot 2024-06-20 at 8 14 12 PM" src="https://github.com/elastic/elastic-agent/assets/121151420/1909f1d7-0332-42b6-b1f4-2bd435d10332">
 </pre>
 <pre> Privileged
<img width="1497" alt="Screenshot 2024-06-20 at 5 35 33 PM" src="https://github.com/elastic/elastic-agent/assets/121151420/de319745-b924-4743-9b55-13e66ec33058">
 </pre>

- ### Host Overview
 <pre> Unprivileged
<img width="1487" alt="Screenshot 2024-06-20 at 5 36 05 PM" src="https://github.com/elastic/elastic-agent/assets/121151420/027e489b-9934-4097-868a-08fe4e2ae445">
<img width="1498" alt="Screenshot 2024-06-20 at 8 20 28 PM" src="https://github.com/elastic/elastic-agent/assets/121151420/d147ca0f-891b-4af1-a412-0e6da26e4c70">
<img width="1501" alt="Screenshot 2024-06-20 at 8 20 39 PM" src="https://github.com/elastic/elastic-agent/assets/121151420/158c92b7-5210-4dfc-b5c5-cddbbc924f89">

 </pre>
 
 > [!Note]
 > Here we can see that `elastic-user-agent` doesn't have access to all processes. I had a vscode window running.
 > It doesn't show up in unprivileged mode but it shows up with privileged access.

 <pre> Privileged
<img width="1508" alt="Screenshot 2024-06-20 at 8 15 13 PM" src="https://github.com/elastic/elastic-agent/assets/121151420/1a5cb80a-e33f-4856-a080-eb3e8289daa0">
<img width="930" alt="Screenshot 2024-06-20 at 5 36 18 PM" src="https://github.com/elastic/elastic-agent/assets/121151420/904746de-e6ed-424f-9be3-410b31e74380">
<img width="1479" alt="Screenshot 2024-06-20 at 5 36 36 PM" src="https://github.com/elastic/elastic-agent/assets/121151420/0ef197a8-5d83-4b92-af93-797e8df27520">

 </pre>

**Comment by VihasMakwana** [comment]
@ycombinator @blakerouse @pierrehilbert @cmacknz 

There's one particular error showing up.
- `Could not return any performance counter values for \\.\C: .Error: Access is denied.`
- This is triggered [here](https://github.com/elastic/elastic-agent-system-metrics/blob/main/metric/system/diskio/diskstat_windows_helper.go#L98-L104) when it tries to open a volume for later `DeviceIoControl`call.

It tries to open a volume (not the filesystem) to fetch performance counters. As per this [microsoft](https://learn.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-createfilea#physical-disks-and-volumes) doc,

> Direct access to the disk or to a volume is restricted.
> - The caller must have administrative privileges. For more information, see [Running with Special Privileges](https://learn.microsoft.com/en-us/windows/desktop/SecBP/running-with-special-privileges).
> 

We can fix this error by giving administrative privileges to our unprivileged user. But isn't it the very thing we're trying to avoid?

How should we tackle this?

cc: @jlind23

**Comment by VihasMakwana** [comment]
> @ycombinator @blakerouse @pierrehilbert @cmacknz
> 
> There's one particular error showing up after giving all necessary privileges.
> 
> * `Could not return any performance counter values for \\.\C: .Error: Access is denied.`
> * This is triggered [here](https://github.com/elastic/elastic-agent-system-metrics/blob/main/metric/system/diskio/diskstat_windows_helper.go#L98-L104) when it tries to open a volume for later `DeviceIoControl`call.
> 
> It tries to open a volume (not the filesystem) to fetch performance counters. As per this [microsoft](https://learn.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-createfilea#physical-disks-and-volumes) doc,
> 
> > Direct access to the disk or to a volume is restricted.
> > 
> > * The caller must have administrative privileges. For more information, see [Running with Special Privileges](https://learn.microsoft.com/en-us/windows/desktop/SecBP/running-with-special-privileges).
> 
> We can fix this error by giving administrative privileges to our unprivileged user. But isn't it the very thing we're trying to avoid?
> 
> How should we tackle this?
> 
> cc: @jlind23

**NOTE**
This error is not fatal, it only affect "DIsk Usage" section in "Host Overview" Dashboards.
`system.diskio` data stream is only thing missing.

<img width="1156" alt="Screenshot 2024-06-20 at 8 54 32 PM" src="https://github.com/elastic/elastic-agent/assets/121151420/69c6f170-b72d-49c1-a2e0-169798d8b2f4">
<img width="1149" alt="Screenshot 2024-06-20 at 8 54 43 PM" src="https://github.com/elastic/elastic-agent/assets/121151420/d00d325a-dc63-43d9-939a-58e873cce5b5">

**Comment by ycombinator** [comment]
> @ycombinator @pierrehilbert now that @VihasMakwana and @kaanyalti tested all the different combinations (unless I am missing something) what would be the next step here? Should @VihasMakwana start writing down all the settings that need to be changed when running unprivileged?

Once @VihasMakwana is done working through all the scenarios (== all checkboxes in the Definition of Done are checked), we will end up with a bunch of comments in this issue going over the scenarios and what changes had to be done to make it work in unprivileged mode.  At that point, we can close this issue. 

@kilfoyle is already aware of this issue and is going to port over the comments into proper documentation.

**Comment by jlind23** [comment]
I would be happy to get @ycombinator and @cmacknz opinion on https://github.com/elastic/elastic-agent/issues/4705#issuecomment-2180986261 but worst case it can be documented as a known limitation.

**Comment by ycombinator** [comment]
> I would be happy to get @ycombinator and @cmacknz opinion on https://github.com/elastic/elastic-agent/issues/4705#issuecomment-2180986261 but worst case it can be documented as a known limitation.

> There's one particular error showing up after giving all necessary privileges.
> `Could not return any performance counter values for \\.\C: .Error: Access is denied.`

I assume this happens when you use the `system` integration on Windows?  Is it for any specific dataset with the `system` integration?

Where are you seeing this error — in the Agent logs? Or somewhere else? Does it occur once or frequently?  Also, are there any other, more visible, symptoms as a result, e.g. an empty dashboard or some message in the Fleet UI?

Where I'm going with these questions is: maybe if Agent knows it's running in unprivileged mode, it could perhaps not even try to access the volume rather than emitting the error (especially if we're emitting this error frequently).  I realize the code for accessing the volume is buried quite deep so this would mean passing down the necessary information to that level.

> We can fix this error by giving administrative privileges to our unprivileged user. But isn't it the very thing we're trying to avoid?

Indeed.  In this case, the "fix" is to run Elastic Agent in privileged mode.  So I would definitely document the symptoms and mention that the observed behavior is expected in unprivileged mode.

**Comment by VihasMakwana** [comment]
> I assume this happens when you use the `system` integration on Windows? Is it for any specific dataset with the `system` integration?
> 

This specifically happens for `system.diskio` dataset.
> Where are you seeing this error — in the Agent logs? Or somewhere else? Does it occur once or frequently? Also, are there any other, more visible, symptoms as a result, e.g. an empty dashboard or some message in the Fleet UI?
> 
Yes, this is in agent logs. It occurs frequently and the frequency depends on the `period` config.
There is an empty section in "Host Overview" dashboard [here](https://github.com/elastic/elastic-agent/issues/4705#issuecomment-2180986261).

> Where I'm going with these questions is: maybe if Agent knows it's running in unprivileged mode, it could perhaps not even try to access the volume rather than emitting the error (especially if we're emitting this error frequently). I realize the code for accessing the volume is buried quite deep so this would mean passing down the necessary information to that level.
> 
I'm doing research on this part. I'll open a separate issue to track this scenario.

**Comment by kilfoyle** [comment]
Thanks @kaanyalti and @VihasMakwana for the super clear guidance around these limitations!

@ycombinator I've tried to capture everything as part of the [Add steps and details for running 'unprivileged' Elastic Agent](https://github.com/elastic/ingest-docs/pull/1087) PR. A preview of the limitations is available in the [Agent and dashboard behaviors in unprivileged mode](https://ingest-docs_bk_1087.docs-preview.app.elstc.co/guide/en/fleet/master/elastic-agent-unprivileged.html#unprivileged-command-behaviors) section of the new page.

Please let me know whatever may need fixing up. :-)
